crates/host/src/capability/mod.rs

```diff
@@ -0,0 +1,216 @@
+#[allow(missing_docs)]
+mod bindgen {
+    wasmtime::component::bindgen!({
+        world: "wasmcloud.wasmcloud",
+        path: "../../wit",
+        async: true,
+    });
+}
+
+pub use bindgen::*;
+
+use logging::Host;
+use rand::{thread_rng, Rng, RngCore};
+
+use core::fmt::Debug;
+
+use std::sync::Arc;
+
+use anyhow::{bail, Context, Result};
+use async_trait::async_trait;
+use futures::lock::Mutex;
+use tracing::instrument;
+
+fn format_opt<T>(opt: &Option<T>) -> &'static str {
+    if opt.is_some() {
+        "set"
+    } else {
+        "unset"
+    }
+}
+
+#[derive(Clone, Default)]
+pub(crate) struct Handler {
+    host: Option<Arc<Mutex<dyn host::Host + Sync + Send>>>,
+    logging: Option<Arc<Mutex<dyn logging::Host + Sync + Send>>>,
+}
+
+impl Debug for Handler {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Handler")
+            .field("host", &format_opt(&self.host))
+            .field("logging", &format_opt(&self.logging))
+            .finish()
+    }
+}
+
+impl Handler {
+    /// Writes a log to configured [`log::Host`]
+    #[instrument]
+    async fn write_log(&mut self, level: &str, text: String) -> anyhow::Result<()> {
+        let level = match level {
+            // NOTE: Trace level is currently missing from wasmbus logging protocol
```

Comment:
Interesting, I think this might've happened because there's no trace logging in Elixir, but looks like this code will work once we add trace